// contentful-migration/built/lib/entities/entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
class Entry {
    constructor(entry) {
        this._id = entry.sys.id;
        this._fields = entry.fields;
        this._version = entry.sys.version;
        this._contentTypeId = entry.sys.contentType.sys.id;
    }
    get id() {
        return this._id;
    }
    get contentTypeId() {
        return this._contentTypeId;
    }
    get fields() {
        return this._fields;
    }
    set fields(fields) {
        this._fields = fields;
    }
    setField(id, value) {
        this._fields[id] = value;
    }
    setFieldForLocale(id, locale, value) {
        const field = this._fields[id] || {};
        field[locale] = value;
        this._fields[id] = field;
    }
    get version() {
        return this._version;
    }
    set version(version) {
        this._version = version;
    }
    toApiEntry() {
        const sys = {
            id: this.id,
            version: this.version,
            contentType: {
                sys: {
                    type: 'Link',
                    linkType: 'ContentType',
                    id: this.contentTypeId
                }
            }
        };
        return {
            sys,
            fields: lodash_1.cloneDeep(this.fields)
        };
    }
    clone() {
        return new Entry(this.toApiEntry());
    }
}
exports.default = Entry;
exports.Entry = Entry;
